src/Lorentz/Contracts/BaseDAO/Proposal.hs

```diff
@@ -727,3 +735,93 @@ flush config = do
 
   iter (handleProposalIsOver config)
   swap; pair
+
+deleteProposals
+  :: forall pm s. NiceParameter pm
+  => Natural : Map Timestamp (ProposalKey pm) : s
+  :-> Map Timestamp (ProposalKey pm)  : s
+deleteProposals = do
+  toNamed #deleteAmount
+  swap; dup; 
+  dip $ do
+    swap
+    push 0; toNamed #counter
+
+  stackType @(Map Timestamp (ProposalKey pm) : ("counter" :! Natural) : ("deleteAmount" :! Natural) : Map Timestamp (ProposalKey pm) : s)
+
+  iter $ do
+    dig @2; dig @2;
+    dupTop2
+    if #counter <. #deleteAmount then do
+      stackType @(("counter" :! Natural) : ("deleteAmount" :! Natural) : (Timestamp, (ProposalKey pm)) : Map Timestamp (ProposalKey pm) : s)
+      -- Delete the proposal
+      dig @3;
+      dig @3;
+      car; dip none; update
+      swap; fromNamed #counter; push (1 :: Natural); add; toNamed #counter
+      dip swap
+    else do
+      -- Increase counter, do nothing
+      dip $ dip drop
+      fromNamed #counter; push (1 :: Natural); add; toNamed #counter
+
+  drop; drop
```

Comment:
Yes, currently this is a batch delete. Before it was affecting ongoing proposals, but now this should only delete finished proposals. 

Since the goal was simply to remove stuck proposals that fail due to decisionLambda, we probably will use this entrypoint to drop the first one, eg. `dropProposals 1`.

I agree that passing `ProposalKey` is better, so if you think we should really do it, I'll make the change. 

cc @gromakovsky

Reply:
Probably my inial proposal about dropping the head proposal was not good. Though let's really ask @gromakovsky.

Reply:
What if we pass only one `ProposalKey`, compare it with the head one and remove the head one if they match (failing otherwise)? It should simplify the contract and make it smaller without significantly complicating its usage.

Reply:
> What if we pass only one `ProposalKey`, compare it with the head one and remove the head one if they match (failing otherwise)? It should simplify the contract and make it smaller without significantly complicating its usage.

I'm fine with this as well. If we're good, I'll make the changes.

Reply:
I'm thinking that in some scenario where we know that the proposal is going to be accepted, but containing fail lambda, we probably want to drop it beforehand to let the proposer fixed the lambda, and avoiding wasting time till the voting period ended. 

Being able to drop only the ended proposals could be limiting. 

(Tho, this should be rarely since, accidentality write a fail lambda should not happen regularly)

Reply:
> I'm thinking that in some scenario where we know that the proposal is going to be accepted, but containing fail lambda, we probably want to drop it beforehand to let the proposer fixed the lambda, and avoiding wasting time till the voting period ended.

Below there is a question about unfreezing tokens in case of failing lambda, I think it's relevant here. There I assumed that we drop only accepted proposals and thus it makes sense to unfreeze all frozen tokens. However, if we can drop potentially negative proposals as well, we may want to reconsider unfreezing.

I think if it's known in advance that the proposal will definitely fail, it simply won't get enough positive votes to pass the quorum threshold and won't be accepted. At least with good UI it sounds like a reasonable assumption.

Reply:
> probably this does not increase safety

I think requiring the admin to explicitly specify what proposal they are removing is a good idea for safety. Or maybe show a list of proposals and "DROP IT" button near the oldest one. Without a good UI it can be easy to make a mistake, especially if there are 2 proposals in the same block. Another case when it can increase safety is when there are multiple people/computers with admin access who can potentially call `dropProposals` at the same time.

> now when we represent the proposals queue as Map, checking whether the given element matches the head one seems expensive

I didn't review code in this PR closely, but if you think it's expensive, we can assume that the administrator will use a good UI (we can write a CLI client that would take `proposalKey` and check it somehow) and won't be accessed by more than 1 person concurrently.

Reply:
If we will just remove a proposal by key, without checking it is at the head, this won't be less safe, or not? At least, concurrency issues do not seem to be possible this way.

Also, generally, being able to remove non-head proposals seems to make sense, since if we `flush` by batches, then some intermediate element may turn out to be failing, and we may want to remove it immediately. Though with the current definitions detecting which exactly element is failing may be difficult (the client may need to do `for [n .. 1] dryRunFlush` to find the first failing proposal).

Reply:
> the client may need to do for [n .. 1] dryRunFlush to find the first failing proposal

Humanity has invented binary search :trollface:

Reply:
I assumed a smart `for`, even more difficult from implementation point of view :smile_cat: